fpx/src/api/models/otel.rs

```diff
@@ -0,0 +1,279 @@
+            span::SpanKind::Producer => SpanKind::Producer,
+            span::SpanKind::Consumer => SpanKind::Consumer,
+        }
+    }
+}
+
+impl From<SpanKind> for libsql::Value {
+    fn from(value: SpanKind) -> Self {
+        value.as_ref().into()
+    }
+}
+
+#[derive(Deserialize, Serialize)]
+pub struct SpanEvent {
+    pub name: String,
+
+    #[serde(with = "time::serde::rfc3339")]
+    pub timestamp: time::OffsetDateTime,
+
+    pub attributes: AttributeMap,
+}
+
+impl From<Event> for SpanEvent {
+    fn from(event: Event) -> Self {
+        Self {
+            name: event.name,
+            timestamp: parse_time_nanos(event.time_unix_nano),
+            attributes: event.attributes.into(),
+        }
+    }
+}
+
+#[derive(Deserialize, Serialize)]
+pub struct SpanLink {
+    pub trace_id: String,
+    pub span_id: String,
+    pub trace_state: String,
+    pub attributes: AttributeMap,
+    pub flags: u32,
+}
+
+impl From<Link> for SpanLink {
+    fn from(link: Link) -> Self {
+        Self {
+            trace_id: hex::encode(link.trace_id),
+            span_id: hex::encode(link.span_id),
+            trace_state: link.trace_state,
+            attributes: link.attributes.into(),
+            flags: link.flags,
+        }
+    }
+}
+
+#[derive(Deserialize, Serialize)]
+pub struct SpanStatus {
+    pub code: SpanStatusCode,
+    pub message: String,
+}
+
+impl From<Status> for SpanStatus {
+    fn from(value: Status) -> Self {
+        Self {
+            code: value.code().into(),
+            message: value.message,
+        }
+    }
+}
+
+#[derive(Deserialize, Serialize)]
+pub enum SpanStatusCode {
+    Unset,
+    Ok,
+    Error,
+}
+
+impl From<StatusCode> for SpanStatusCode {
+    fn from(status_code: StatusCode) -> Self {
+        match status_code {
+            StatusCode::Unset => SpanStatusCode::Unset,
+            StatusCode::Ok => SpanStatusCode::Ok,
+            StatusCode::Error => SpanStatusCode::Error,
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone, Default)]
+pub struct AttributeMap(BTreeMap<String, Option<AttributeValue>>);
+
+impl From<KeyValueList> for AttributeMap {
+    fn from(value: KeyValueList) -> Self {
+        value.values.into()
+    }
+}
+
+impl From<Vec<KeyValue>> for AttributeMap {
+    fn from(attributes: Vec<KeyValue>) -> Self {
+        let result: BTreeMap<String, Option<AttributeValue>> = attributes
+            .into_iter()
+            .map(|kv| {
+                (
+                    kv.key,
+                    kv.value.and_then(|value| value.value.map(Into::into)),
+                )
+            })
+            .collect();
+
+        AttributeMap(result)
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone)]
+pub enum AttributeValue {
+    String(String),
+    Bool(bool),
+    Int(i64),
+    Double(f64),
+    Array(Vec<AttributeValue>),
+    KeyValueList(AttributeMap),
+    Bytes(Vec<u8>),
+}
+
+impl From<any_value::Value> for AttributeValue {
```

Comment:
why do this instead of using `dyn Any` with `.downcast_ref`?

Reply:
Not sure I understand you suggestion. The current reference is related to a type within the opentelemetry-otel to our type. The `Any` is a Rust trait, right?